src/tamp_improv/approaches/rl_improvisational_policy.py

```diff
@@ -0,0 +1,134 @@
+
+
+class TrainingProgressCallback(BaseCallback):
+    """Tracks and reports training progress.
+
+    This gets called after each training step to monitor:
+    - Success rate (how often we successfully clear the target area)
+    - Episode lengths (how many steps it takes to succeed)
+    """
+
+    def __init__(self, check_freq: int = 1000, verbose: int = 1):
+        super().__init__(verbose)
+        self.check_freq = check_freq
+        self.success_history: list[bool] = []
+        self.episode_lengths: list[int] = []
+        self.current_length = 0  # Counter for current episode
+
+    def _on_step(self) -> bool:
+        """Called after each training step."""
+        self.current_length += 1
+        # dones indicates if episode ended this step
+        dones = self.locals["dones"]
+        infos = self.locals["infos"]
+
+        if dones[0]:  # If episode ended
+            # Check if it ended due to success (clearing area) vs timeout
+            success = not infos[0].get("TimeLimit.truncated", False)
+            self.success_history.append(success)
+            self.episode_lengths.append(self.current_length)
+            self.current_length = 0
+
+            # Print stats periodically
+            if len(self.success_history) % self.check_freq == 0:
+                recent_successes = self.success_history[-self.check_freq :]
+                recent_lengths = self.episode_lengths[-self.check_freq :]
+                success_rate = sum(recent_successes) / len(recent_successes)
+                avg_length = sum(recent_lengths) / len(recent_lengths)
+                print(f"Episodes: {len(self.success_history)}")
+                print(f"Success rate: {success_rate:.2%}")
+                print(f"Average episode length: {avg_length:.1f}")
+
+        return True  # Continue training
+
+
+class RLImprovisationalPolicy(
+    ImprovisationalPolicy[NDArray[np.float32], NDArray[np.float32]]
+):
+    """The actual RL policy that learns to push block 2 away."""
+
+    def __init__(self, env: PushingEnvWrapper):
+        """Initialize policy with a PPO model.
+
+        PPO hyperparameters:
+        - learning_rate: How fast to update the policy
+        - n_steps: How many steps to collect before updating
+        - batch_size: How many samples to use in each update
+        - n_epochs: How many times to reuse each sample
+        - gamma: Discount factor for future rewards
+        """
+        if not isinstance(env, PushingEnvWrapper):
+            raise TypeError("Environment must be a PushingEnvWrapper instance")
+
+        self.env = env
+        # Create PPO model with the environment
+        self.model = PPO(
+            "MlpPolicy",
+            self.env,
+            learning_rate=3e-4,
+            n_steps=2048,
+            batch_size=64,
+            n_epochs=10,
+            gamma=0.99,
+            verbose=1,
+        )
+
+    def train(
+        self, total_timesteps: int = 1_000_000, seed: Optional[int] = None
```

Comment:
woah that's super cool, I didn't know that you could have underscores in numbers like that! thanks for teaching me :)